File: ios/Classes/DeviceLocatorBridge.swift
import Flutter
import UIKit

// `FlutterError` inherits from NSObject, not NSError, so it does not automatically
// conform to Swift's `Error` protocol.  This retroactive conformance lets us
// `throw FlutterError(...)` from any method that declares `throws`.
extension FlutterError: @retroactive Error {}

// MARK: - C error helper

/// Converts a `clCError` to a thrown `FlutterError` when the operation failed.
/// The `error.message` field is a fixed-size C char array (`char[256]`); read it
/// via `withUnsafePointer` + `withMemoryRebound` to obtain a Swift `String`.
func checkCError(_ error: clCError) throws {
    guard error.success else {
        let message = withUnsafePointer(to: error.message) {
            $0.withMemoryRebound(to: CChar.self, capacity: 256) {
                String(cString: $0)
            }
        }
        throw FlutterError(
            code: String(error.code.rawValue),
            message: message,
            details: nil
        )
    }
}

// MARK: - DeviceLocatorBridge

class DeviceLocatorBridge: NSObject, FlutterStreamHandler {

    // MARK: Static state (used by C callbacks which cannot capture `self`)

    /// The currently active bridge. C callbacks reach the instance through this
    /// static weak reference because `clCDeviceLocator_SetOnDeviceListEvent`
    /// takes a bare C function pointer with no `void* context` parameter.
    private static weak var activeBridge: DeviceLocatorBridge?

    /// Device handles created by `createDevice`, keyed by serial number.
    /// `DeviceBridge` looks up entries here when it needs an existing handle.
    static var devices: [String: OpaquePointer] = [:]

    // MARK: Instance state

    private var locator: OpaquePointer?
    private var deviceListSink: FlutterEventSink?

    // MARK: - Lifecycle methods

    /// Creates the native `clCDeviceLocator`. Optionally writes SDK logs to `logDirectory`.
    func create(logDirectory: String?) throws {
        var error = clCError()
        if let dir = logDirectory {
            locator = clCDeviceLocator_CreateWithLogDirectory(dir, &error)
        } else {
            locator = clCDeviceLocator_Create(&error)
        }
        try checkCError(error)
    }

    /// Creates a `clCDevice` handle for `serial` and stores it in `DeviceLocatorBridge.devices`.
    /// Returns `nil` — the Dart side constructs the `Device` object locally from the serial.
    @discardableResult
    func createDevice(serial: String) throws -> Any? {
        guard let loc = locator else {
            throw FlutterError(code: "NO_LOCATOR", message: "DeviceLocator not created", details: nil)
        }
        var error = clCError()
        let device = clCDeviceLocator_CreateDevice(loc, serial, &error)
        try checkCError(error)
        if let device = device {
            DeviceLocatorBridge.devices[serial] = device
        }
        return nil
    }

    /// Switches the SDK between multi-threaded (default) and single-threaded mode.
    /// In single-threaded mode `update()` must be called periodically from the main thread.
    func setSingleThreaded(enabled: Bool) {
        clCCapsule_SetSingleThreaded(enabled)
    }

    /// Pumps pending SDK callbacks when operating in single-threaded mode.
    func update() throws {
        guard let loc = locator else {
            throw FlutterError(code: "NO_LOCATOR", message: "DeviceLocator not created", details: nil)
        }
        clCDeviceLocator_Update(loc)
    }

    /// Sets the SDK log verbosity level.
    func setLogLevel(level: Int) {
        clCCapsule_SetLogLevel(clCCapsule_LogLevel(rawValue: UInt32(level)))
    }

    /// Releases the native locator handle and clears all stored device handles.
    func dispose() {
        if let loc = locator {
            clCDeviceLocator_Destroy(loc)
            locator = nil
        }
        DeviceLocatorBridge.devices.removeAll()
        if DeviceLocatorBridge.activeBridge === self {
            DeviceLocatorBridge.activeBridge = nil
        }
    }

    deinit {
        if let loc = locator {
            clCDeviceLocator_Destroy(loc)
        }
        DeviceLocatorBridge.devices.removeAll()
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        guard let loc = locator else {
            return FlutterError(code: "NO_LOCATOR", message: "DeviceLocator not created", details: nil)
        }
        guard let args = arguments as? [String: Any],
              let deviceType = args["deviceType"] as? Int,
              let searchTime = args["searchTime"] as? Int else {
            return FlutterError(
                code: "INVALID_ARGS",
                message: "Missing 'deviceType' or 'searchTime'",
                details: nil
            )
        }

        deviceListSink = events
        DeviceLocatorBridge.activeBridge = self

        // Register a non-capturing @convention(c) callback. The callback reaches
        // the bridge instance via the static `activeBridge` weak reference.
        clCDeviceLocator_SetOnDeviceListEvent(loc) { _, list, failReason in
            guard let bridge = DeviceLocatorBridge.activeBridge,
                  let sink = bridge.deviceListSink else { return }

            if failReason != clCDeviceLocator_FailReason_OK {
                let flutterError: FlutterError
                if failReason == clCDeviceLocator_FailReason_BluetoothDisabled {
                    // code "1" = NeiryErrorCode.failedToConnect → Dart: BluetoothDisabledException
                    flutterError = FlutterError(
                        code: "1",
                        message: "Bluetooth is disabled",
                        details: nil
                    )
                } else {
                    // code "255" = NeiryErrorCode.unknown
                    flutterError = FlutterError(
                        code: "255",
                        message: "Unknown error during device search",
                        details: nil
                    )
                }
                DispatchQueue.main.async {
                    sink(flutterError)
                    sink(FlutterEndOfEventStream)
                }
                return
            }

            // Build the device-info array from the list.
            var devicesArray: [[String: Any]] = []
            if let list = list {
                var countError = clCError()
                let count = clCDeviceInfoList_GetCount(list, &countError)
                if countError.success {
                    for i in 0..<count {
                        var infoError = clCError()
                        if let info = clCDeviceInfoList_GetDeviceInfo(list, i, &infoError),
                           infoError.success {
                            let serial = String(cString: clCDeviceInfo_GetSerial(info))
                            let name   = String(cString: clCDeviceInfo_GetName(info))
                            // .rawValue converts clCDeviceType C enum to Int expected by
                            // DeviceInfo.fromMap on the Dart side.
                            let type   = Int(clCDeviceInfo_GetType(info).rawValue)
                            devicesArray.append(["serial": serial, "name": name, "type": type])
                        }
                    }
                }
            }

            DispatchQueue.main.async {
                sink(devicesArray)
                sink(FlutterEndOfEventStream)
            }
        }

        // Start the BLE scan — callbacks fire asynchronously when the timeout expires.
        var requestError = clCError()
        clCDeviceLocator_RequestDevices(
            loc,
            clCDeviceType(rawValue: UInt32(deviceType)),
            Int32(searchTime),
            &requestError
        )
        if !requestError.success {
            let message = withUnsafePointer(to: requestError.message) {
                $0.withMemoryRebound(to: CChar.self, capacity: 256) {
                    String(cString: $0)
                }
            }
            let err = FlutterError(
                code: String(requestError.code.rawValue),
                message: message,
                details: nil
            )
            deviceListSink = nil
            DeviceLocatorBridge.activeBridge = nil
            events(err)
            events(FlutterEndOfEventStream)
        }

        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        deviceListSink = nil
        if DeviceLocatorBridge.activeBridge === self {
            DeviceLocatorBridge.activeBridge = nil
        }
        // Clearing deviceListSink is sufficient — pending callbacks will early-return
        // via the `guard let sink = bridge.deviceListSink` check above.
        return nil
    }
}

File: android/src/main/kotlin/com/neiry/neiry_kit/DeviceLocatorBridge.kt
package com.neiry.neiry_kit

import android.os.Handler
import android.util.Log
import io.flutter.plugin.common.EventChannel
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Kotlin bridge for `clCDeviceLocator_*`. Owns the native locator handle and
 * implements [EventChannel.StreamHandler] for the `deviceList` event stream.
 *
 * Mirrors [DeviceLocatorBridge.swift] in behavior.
 */
class DeviceLocatorBridge(
    private val nativeBridge: NativeBridge,
    private val mainHandler: Handler,
) : EventChannel.StreamHandler {

    private var handle: Long = 0L

    /**
     * The currently active guarded sink. Guards two classes of spurious endOfStream:
     *
     * Class 1 — stale from previous scan:
     * SDK queues endOfStream via Handler.post during scan A; user starts scan B before
     * that fires; queued lambda fires on scan B's Dart handler → "No element".
     * Guard: identity check `currentSink === this` — replaced on each onListen.
     *
     * Class 2 — deferred SDK reset (~54ms after nativeRequestDevices):
     * Native C SDK fires endOfStream on the current scan's sink ~54ms into a new scan,
     * resetting leftover state from the previous scan. No success or error precedes it.
     * Guard: per-sink `eventReceived` flag — only allow endOfStream through if at least
     * one success or error event arrived first. A scan that closes with no data is
     * definitionally a spurious SDK reset.
     */
    @Volatile private var currentSink: EventChannel.EventSink? = null

    companion object {
        /** Device handles keyed by serial. DeviceBridge looks up entries here. */
        var devices: MutableMap<String, Long> = mutableMapOf()
    }

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    /** Creates the native `clCDeviceLocator`. */
    fun create(logDirectory: String?) {
        try {
            handle = nativeBridge.nativeCreateLocator(logDirectory)
        } catch (e: RuntimeException) {
            throw parseSdkError(e.message ?: "255|Unknown error")
        }
    }

    /**
     * Creates a device handle for [serial] and stores it in [devices].
     * Returns the native handle (stored internally; Dart does not need it).
     */
    fun createDevice(serial: String): Long {
        if (handle == 0L) {
            throw FlutterError("NO_LOCATOR", "DeviceLocator not created", null)
        }
        return try {
            val deviceHandle = nativeBridge.nativeCreateDevice(handle, serial)
            devices[serial] = deviceHandle
            deviceHandle
        } catch (e: RuntimeException) {
            throw parseSdkError(e.message ?: "255|Unknown error")
        }
    }

    /** Switches the SDK between multi-threaded and single-threaded mode. */
    fun setSingleThreaded(enabled: Boolean) {
        nativeBridge.nativeSetSingleThreaded(enabled)
    }

    /** Pumps SDK callbacks when in single-threaded mode. */
    fun update() {
        if (handle == 0L) {
            throw FlutterError("NO_LOCATOR", "DeviceLocator not created", null)
        }
        nativeBridge.nativeUpdate(handle)
    }

    /** Sets SDK log verbosity. */
    fun setLogLevel(level: Int) {
        nativeBridge.nativeSetLogLevel(level)
    }

    /** Releases the native locator and clears all device handles. */
    fun dispose() {
        if (handle != 0L) {
            nativeBridge.nativeSetDeviceListSink(null)
            nativeBridge.nativeDestroyLocator(handle)
            handle = 0L
        }
        devices.clear()
    }

    // ── EventChannel.StreamHandler ────────────────────────────────────────────

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        if (handle == 0L) {
            events?.error("NO_LOCATOR", "DeviceLocator not created", null)
            events?.endOfStream()
            return
        }

        val args       = arguments as? Map<*, *>
        val deviceType = args?.get("deviceType") as? Int
        val searchTime = args?.get("searchTime") as? Int

        if (deviceType == null || searchTime == null) {
            events?.error("INVALID_ARGS", "Missing 'deviceType' or 'searchTime'", null)
            events?.endOfStream()
            return
        }

        // eventReceived starts false; only success or error sets it true.
        // endOfStream is only forwarded to Dart if eventReceived is true — a scan
        // that closes with no prior data is a spurious SDK reset and is silently dropped.
        val eventReceived = AtomicBoolean(false)

        val wrappedSink = events?.let { realSink ->
            object : EventChannel.EventSink {
                override fun success(event: Any?) {
                    if (currentSink === this) {
                        eventReceived.set(true)
                        realSink.success(event)
                    }
                }
                override fun error(code: String, message: String?, details: Any?) {
                    if (currentSink === this) {
                        eventReceived.set(true)
                        realSink.error(code, message, details)
                    }
                }
                override fun endOfStream() {
                    val isCurrent = currentSink === this
                    val hasEvent  = eventReceived.get()
                    Log.d("Neiry", "DeviceLocatorBridge.endOfStream isCurrent=$isCurrent hasEvent=$hasEvent")
                    if (isCurrent && hasEvent) realSink.endOfStream()
                }
            }
        }
        currentSink = wrappedSink

        // Register the sink BEFORE starting the scan so the callback always finds it set.
        nativeBridge.nativeSetDeviceListSink(wrappedSink)

        try {
            Log.d("Neiry", "DeviceLocatorBridge.onListen: nativeRequestDevices")
            nativeBridge.nativeRequestDevices(handle, deviceType, searchTime)
        } catch (e: RuntimeException) {
            val err = parseSdkError(e.message ?: "255|Unknown error")
            currentSink = null
            events?.error(err.code, err.message, err.details)
            events?.endOfStream()
            nativeBridge.nativeSetDeviceListSink(null)
            return
        }
    }

    override fun onCancel(arguments: Any?) {
        Log.d("Neiry", "DeviceLocatorBridge.onCancel")
        currentSink = null
        nativeBridge.nativeSetDeviceListSink(null)
    }
}
